Add ns1_record module and tests

mburtless opened the pull request:
##### SUMMARY
NS1 (https://ns1.com/) would like to contribute an official module for working with resources in our managed DNS platform to Ansible.  The intent of this PR is to establish the necessary namespace, module utils, documentation fragments, and the first module, ns1_record.  This module maintains DNS records in existing DNS zones that are hosted with NS1.

This PR also contains the necessary changes to enable integration testing, a cloud config template (as an API key is required) and integration tests for the first module.

Following this PR we have several other modules we would like to add, including ns1_zone to manage hosted DNS zones and ns1_facts to gather facts about resources hosted with NS1.

##### ISSUE TYPE
- New Module Pull Request

##### COMPONENT NAME
ns1_record

##### ADDITIONAL INFORMATION
The NS1 python SDK, ns1-python, is required for testing and using this module.  This can be installed via pip (`pip install ns1-python`).  An NS1 API key must be defined in `test/integration/cloud-config-ns1.ini` in order to run integration tests against the NS1 platform.  A key can be obtained by signing up for a free NS1 developer account at https://ns1.com/signup

mburtless committed:
Add ns1_record module and tests

mattclay commented on lib/ansible/module_utils/ns1.py:
```diff
@@ -0,0 +1,55 @@
+try:
+    from ansible.module_utils.ansible_release import __version__ as ANSIBLE_VERSION
+except Exception:
+    ANSIBLE_VERSION = 'unknown'
```
This import isn't used and should be removed.

mburtless replied:
Thanks!  Meant to use this in user agent string, but didn't get around to implementing this.

resmo commented on lib/ansible/module_utils/ns1.py:
```diff
@@ -0,0 +1,55 @@
+# -*- coding: utf-8 -*-
+
+# Copyright: (c) 2019, Matthew Burtless <mburtless@ns1.com>
+# GNU General Public License v3.0+ (see COPYING or https://www.gnu.org/licenses/gpl-3.0.txt)
```
module utils should be bsd licensed

mburtless replied:
Thanks for the catch, missed that requirement from the docs

mburtless committed:
Adds fixes for sanity test and review results

- Fix dictionary comprehension format to be py2.6 compatable
- Fix whitespace error
- Fix version number in version_added
- Removes unused import per review by mattclay
- Modifies lisence in module util to BSD per review by resmo

mburtless commented:
Issues from previous sanity tests and reviews by @resmo and @mattclay should now be resolved.

Shippable tests seemed to fail with no error after a timeout.  Console output seemed to be hung on a step unrelated to this module (`Create user with ssh key`).  Can anyone point me in the right direction to ensure CI tests are passing as expected?

mburtless commented:
At this point all issues from sanity tests and reviews have been resolved.  Can anyone comment on what the next step is here?

mattclay commented on test/integration/targets/ns1_record/aliases:
```diff
@@ -0,0 +1,3 @@
+shippable/cloud/group1
```
Use `unsupported` here instead, since these tests cannot run on Shippable.

mburtless replied:
Makes sense, thanks.  Should be all set.

mburtless committed:
Mark tests unsupported and remove from Shippable group

mburtless committed:
Update integration tests to target randomly generated domain name

mattclay commented:
bot_status

mburtless committed:
Bump version_added

mburtless commented:
Hi @mattclay - Is there anything outstanding that needs to be completed before this can be merged?

igor-raits commented on lib/ansible/plugins/doc_fragments/ns1.py:
```diff
@@ -0,0 +1,25 @@
+
+requirements:
+  - python >= 2.7
+  - ns1 >= 0.9.19
```
`ns1-python`?

mburtless replied:
The name in PyPi is `ns1-python` but the namespace is `ns1`.  Wasn't clear to me which we should be listing here (or if it matters), but I think you're right that package name probably makes more sense.

igor-raits replied:
true.. I think PyPI package name makes more sense.

mburtless committed:
Update requirement name to package name

ignatenkobrain commented:
@mburtless check CI btw, it fails on some documentation stuff

mburtless committed:
Fixes for stricter boilerplate and documentation sanity tests

mburtless commented:
Hi @mattclay @resmo - Checking in again, I'd love to hear some feedback or get this merged, if possible.

Is there anything outstanding that needs to be completed before this can be merged?

mattclay commented on test/runner/lib/cloud/ns1.py:
```diff
@@ -0,0 +1,57 @@
+"""NS1 plugin for integration tests."""
```
These plugins have moved to `test/lib/ansible_test/_internal/cloud/`

mattclay commented on test/integration/cloud-config-ns1.ini.template:
```diff
@@ -0,0 +1,12 @@
+# This is the configuration template for ansible-test NS1 integration tests.
```
These templates have moved to `test/lib/ansible_test/config/`

mburtless committed:
Update integration test file paths

mburtless committed:
Remove legacy test file

mburtless committed:
Update import paths

mburtless commented:
Hi @mattclay - Thanks for the heads up.  I've rebased the branch on devel and moved the integration test files and to their new locations.  Please let me know if anything else is required here.

mburtless commented:
Hi @mattclay @resmo - Checking in once more.  Would love to hear some further feedback or get this merged, if possible.

Is there anything outstanding that needs to be completed before this can be merged?

mburtless commented:
Hi @mattclay @resmo  - Checking in again. Is there anything outstanding that needs to be completed before this can be merged?

mattclay commented:
@mburtless Between getting ready for the Ansible 2.9 release and AnsibleFest, I probably won't have time to look at this for at least 2 weeks.

mburtless commented:
@mattclay - Thanks for the heads up, appreciate the timeline.  To clarify, does that mean that this module and the associated namespace will not make it into 2.9?

mattclay commented:
@mburtless Correct. Freeze for Ansible 2.9 was on August 29th:

https://docs.ansible.com/ansible/latest/roadmap/ROADMAP_2_9.html#expected

gundalow reviewed (commented):
~~Hi, I wonder if the modules should be moved into https://github.com/ansible/ansible/tree/devel/lib/ansible/modules/cloud/digital_ocean

@BondAnthony mgregson fyi~~

Urgh, sorry I miss read and thought this was part of DO

dch commented:
@mattclay any chance of this landing in the next ansible relase? thanks!

Akasurde commented:
Hi @mburtless, Thank you very much for your interest in Ansible. This plugin/module is no longer maintained in this repository and has been migrated to https://github.com/ansible-collections/community.general<br><br>If you have further questions please stop by IRC or the mailing list:<br><br>* IRC: #ansible on irc.freenode.net<br>* mailing list: https://groups.google.com/forum/#!forum/ansible-project<br><br>needs_info<br>
